Source/WebGPU/WebGPU/CommandEncoder.mm

```diff
@@ -352,6 +352,69 @@ static auto timestampWriteIndex(auto writeIndex)
     return std::make_pair(pso, depthStencil);
 }
 
+void CommandEncoder::runDepthClampEncoder(float minDepth, float maxDepth, NSMutableDictionary<NSNumber*, TextureAndClearColor*> *colorAttachments, id<MTLTexture> depthStencilTexture, id<MTLRenderCommandEncoder> renderCommandEncoder)
+{
+    if (minDepth <= 0.f || maxDepth >= 1.f || !depthStencilTexture)
+        return;
+
+    auto pixelFormat = depthStencilTexture.pixelFormat;
+    if (pixelFormat == MTLPixelFormatStencil8 || pixelFormat == MTLPixelFormatX32_Stencil8)
+        return;
+
+    NSError *error = nil;
+    MTLCompileOptions* options = [MTLCompileOptions new];
+ALLOW_DEPRECATED_DECLARATIONS_BEGIN
+    options.fastMathEnabled = YES;
+ALLOW_DEPRECATED_DECLARATIONS_END
+    id<MTLDevice> device = m_device->device();
+    id<MTLLibrary> library = [device newLibraryWithSource:[NSString stringWithFormat:@"using namespace metal;\n[[vertex]] float4 vsMin(unsigned int vid [[vertex_id]]) { float d = %f; float4 v[] = { float4(-1,-1,d,1), float4(-1,4,d,1), float4(4,-1,d,1) }; return v[vid]; }\n[[vertex]] float4 vsMax(unsigned int vid [[vertex_id]]) { float d = %f; float4 v[] = { float4(-1,-1,d,1), float4(-1,4,d,1), float4(4,-1,d,1) }; return v[vid]; }\n struct S { float d [[depth(any)]]; }; \n[[fragment]] S fsMin()\n{ return S { .d = %f }; }\n[[fragment]] S fsMax()\n{ return S { .d = %f }; }", minDepth, maxDepth, minDepth, maxDepth] options:options error:&error];
```

Comment:
Could we pull this source string out into a static constant R"()" string literal to make it easier to read?

What's the significance of the 4 in `float4(-1, 4, f, 1)`? And does `vid` need to be modulo 3?

Reply:
I only draw 1 triangle.

It's a triangle which covers the entire viewport post transform. Do I have my math wrong (very possible)? The post-transform cube has bounds [-1,1] in x and y directions so I draw a triangle like this:
```
|\
| \
|  \
|   \
|    \
|     \
|      \
|       \
|-------|\
|       | \
|       |  \
|-------|___\
```
let me re-run the math on that square right now

Reply:
oh good catch, the 4 should be a 3

Reply:
in practice it won't matter since the triangle will be a little too large

Reply:
I will make it an `"R("` string good idea.

The idea on the z-component is that only values less than min depth and values greater than max depth should pass the depth test. It is used in combination with the depth comparison functions.

Reply:
@djg so this whole extra pass was completely unnecessary. I added it before I added the WGSL code generation, thinking we could do without extra instructions in the generated WGSL code. However, the CTS revealed it was inadequate and the code-gen was needed.

With the extra codegen, we don't need the extra render pass, so I just removed it.